File: src/domain/risk.ts
/*
 * Centralized risk classification.
 *
 * The README is explicit: "Risk classification (RISK map) recolors avatars,
 * left borders, badges everywhere — centralize it." Every surface (app + portal)
 * imports from here so colors and thresholds never drift.
 */
import type { AncVisit, RiskLevel } from './types';

export interface RiskStyle {
  color: string;
  bg: string;
  label: string;
  bn: string;
}

export const RISK: Record<RiskLevel, RiskStyle> = {
  high: { color: '#C0492E', bg: '#F7E4DE', label: 'High risk', bn: 'উচ্চ ঝুঁকি' },
  medium: { color: '#B5792A', bg: '#F5EAD6', label: 'Medium risk', bn: 'মাঝারি ঝুঁকি' },
  low: { color: '#0E7C66', bg: '#E0EFEA', label: 'Low risk', bn: 'কম ঝুঁকি' },
};

const ORDER: Record<RiskLevel, number> = { low: 0, medium: 1, high: 2 };

/** Returns the more severe of two risk levels. */
export function maxRisk(a: RiskLevel, b: RiskLevel): RiskLevel {
  return ORDER[a] >= ORDER[b] ? a : b;
}

const DANGER_LABELS: Record<keyof AncVisit['danger'], string> = {
  headache: 'Severe headache',
  blurred: 'Blurred vision',
  swelling: 'Swelling of face / hands',
  bleeding: 'Vaginal bleeding',
  convulsions: 'Convulsions',
  fever: 'High fever',
  fetal: 'Reduced fetal movement',
  abdo: 'Severe abdominal pain',
  breathing: 'Difficult breathing',
};

export interface ComputedRisk extends RiskStyle {
  level: RiskLevel;
  reasons: string[];
}

/**
 * Auto-classify an ANC visit from vitals + danger signs.
 * Any danger sign → high. BP / Hb / FHR thresholds bump the level.
 */
export function computeAncRisk(v: AncVisit): ComputedRisk {
  const reasons: string[] = [];
  let level: RiskLevel = 'low';
  const bump = (l: RiskLevel) => {
    level = maxRisk(level, l);
  };

  (Object.keys(v.danger) as (keyof AncVisit['danger'])[]).forEach((k) => {
    if (v.danger[k]) {
      reasons.push(DANGER_LABELS[k]);
      bump('high');
    }
  });

  if (v.sbp >= 140 || v.dbp >= 90) {
    reasons.push(`Raised BP ${v.sbp}/${v.dbp} mmHg (pre-eclampsia risk)`);
    bump('high');
  } else if (v.sbp >= 130 || v.dbp >= 85) {
    reasons.push(`Borderline BP ${v.sbp}/${v.dbp}`);
    bump('medium');
  }

  if (v.hb < 7) {
    reasons.push(`Severe anaemia · Hb ${v.hb} g/dL`);
    bump('high');
  } else if (v.hb < 11) {
    reasons.push(`Anaemia · Hb ${v.hb} g/dL`);
    bump('medium');
  }

  if (v.fhr < 120 || v.fhr > 160) {
    reasons.push(`Abnormal fetal heart rate ${v.fhr} bpm`);
    bump('high');
  }

  return { level, reasons, ...RISK[level] };
}

File: mobile/lib/domain/risk.dart
/// Centralized risk classification. Every surface (field app + portal) reads
/// from here so colours and thresholds never drift — ported from the original
/// `risk.ts` and kept in lock-step with the backend's RiskService.
import 'package:flutter/material.dart';

import 'models.dart';

class RiskStyle {
  const RiskStyle(this.color, this.bg, this.label, this.bn);
  final Color color;
  final Color bg;
  final String label;
  final String bn;
}

const Map<RiskLevel, RiskStyle> kRisk = {
  RiskLevel.high: RiskStyle(Color(0xFFC0492E), Color(0xFFF7E4DE), 'High risk', 'উচ্চ ঝুঁকি'),
  RiskLevel.medium: RiskStyle(Color(0xFFB5792A), Color(0xFFF5EAD6), 'Medium risk', 'মাঝারি ঝুঁকি'),
  RiskLevel.low: RiskStyle(Color(0xFF0E7C66), Color(0xFFE0EFEA), 'Low risk', 'কম ঝুঁকি'),
};

RiskStyle riskStyle(RiskLevel level) => kRisk[level]!;

class ComputedRisk {
  ComputedRisk(this.level, this.reasons) : style = riskStyle(level);
  final RiskLevel level;
  final List<String> reasons;
  final RiskStyle style;

  Color get color => style.color;
  Color get bg => style.bg;
  String get label => style.label;
  String get bn => style.bn;
}

const Map<String, String> _dangerLabels = {
  'headache': 'Severe headache',
  'blurred': 'Blurred vision',
  'swelling': 'Swelling of face / hands',
  'bleeding': 'Vaginal bleeding',
  'convulsions': 'Convulsions',
  'fever': 'High fever',
  'fetal': 'Reduced fetal movement',
  'abdo': 'Severe abdominal pain',
  'breathing': 'Difficult breathing',
};

String _trim(double v) => v == v.roundToDouble() ? v.toInt().toString() : v.toString();

/// Auto-classify an ANC visit from vitals + danger signs.
/// Any danger sign → high. BP / Hb / FHR thresholds bump the level.
ComputedRisk computeAncRisk(AncVisit v) {
  final reasons = <String>[];
  var level = RiskLevel.low;
  void bump(RiskLevel l) => level = RiskLevel.max(level, l);

  final flags = <String, bool>{
    'headache': v.danger.headache,
    'blurred': v.danger.blurred,
    'swelling': v.danger.swelling,
    'bleeding': v.danger.bleeding,
    'convulsions': v.danger.convulsions,
    'fever': v.danger.fever,
    'fetal': v.danger.fetal,
    'abdo': v.danger.abdo,
    'breathing': v.danger.breathing,
  };
  flags.forEach((key, on) {
    if (on) {
      reasons.add(_dangerLabels[key]!);
      bump(RiskLevel.high);
    }
  });

  if (v.sbp >= 140 || v.dbp >= 90) {
    reasons.add('Raised BP ${v.sbp}/${v.dbp} mmHg (pre-eclampsia risk)');
    bump(RiskLevel.high);
  } else if (v.sbp >= 130 || v.dbp >= 85) {
    reasons.add('Borderline BP ${v.sbp}/${v.dbp}');
    bump(RiskLevel.medium);
  }

  if (v.hb < 7) {
    reasons.add('Severe anaemia · Hb ${_trim(v.hb)} g/dL');
    bump(RiskLevel.high);
  } else if (v.hb < 11) {
    reasons.add('Anaemia · Hb ${_trim(v.hb)} g/dL');
    bump(RiskLevel.medium);
  }

  if (v.fhr < 120 || v.fhr > 160) {
    reasons.add('Abnormal fetal heart rate ${v.fhr} bpm');
    bump(RiskLevel.high);
  }

  return ComputedRisk(level, reasons);
}
